### PR TITLE
fix patch strip

### LINKED ISSUE #4453: tools.patch(..., strip=n) does not work with new/deleted files
This seems related to #1807

#1898 has added code in conan to support creating new files, but this code does not adjust the source/target paths per strip=n before calling save/unlink.

conan 1.11.2, windows 7

Specifically, the code here needs to trim `strip` leading components off the path before joining it to base_path:

https://github.com/conan-io/conan/blob/fd0507594838413e09abbf8bb7be128c23dd6378/conans/client/tools/files.py#L206-L215

A simple example based on the `conan new` template is at https://gist.github.com/puetzk/d8553b06827ec4f2071c2428b118bd88
`conan source .` runs, and the second hunk of example.patch is applied to hello/hello.cpp correctly (stripping src), but the first hunk is not stripped and therefore creates src/hello/goodbye.cpp instead of hello/goodbye.cpp

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Bugfix: ``tools.patch`` strip!=0 parameter was not working for created or deleted files.
Docs: omit

Close #4453

### CODE REVIEW

File: conans/test/unittests/tools/files_patch_test.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+@@ -1 +1 @@
+-ONE TWO THREE
++ONE TWO DOH!\'''
+        patch(patch_string=patch_content)
+
+'''
+        tmp_dir, file_path, text_file = self._save_files(file_content)
+        self._build_and_check(tmp_dir, file_path, text_file, "ONE TWO DOH!")
+
+    def test_patch_new(self):
```
Comment (memsharded) [inline conans/test/unittests/tools/files_patch_test.py]: This is the only test that is new, the other tests are just moved from the other location, sorry for mixing both in the same PR.

File: conans/client/tools/files.py [hunk 2]
```diff
@@ -194,24 +194,30 @@ def emit(self, record):
     if not patchset:
         raise ConanException("Failed to parse patch: %s" % (patch_file if patch_file else "string"))
 
+    def decode_clean(path, prefix):
+        path = path.decode("utf-8").replace("\\", "/")
+        if path.startswith(prefix):
+            path = path[2:]
+        return path
+
+    def handle_null(path):
```
Comment (lasote) [inline conans/client/tools/files.py]: Why the name `handle_null`? Shouldn't be something related with the `strip` parameter?

File: conans/test/unittests/tools/files_patch_test.py (hunk 1 above)
Comment (danimtb) [reply conans/test/unittests/tools/files_patch_test.py]: I don't think that tests using the TestClient should be considered as unit testing. I would move this test to functional or do a real unit test of `tools.pathc()` in this case

Comment (memsharded) [reply conans/test/unittests/tools/files_patch_test.py]: These tests were already in ``unittests``, I just gave them a dedicated file according to the layout.
They should be refactored and converted to real unit tests, it is not necessary that they use the TestClient.

In any case, I know it is not perfect, but I am not uncomfortable with some test in ``unittests`` that test something very, very specific like ``tools.patch()`` within a recipe. They shouldn't be the norm, but It is nice to have some complete test, and moving that single test to somewhere in ``functional`` would separate the logic too much.

One of the advantages of this structure of tests is knowing what to run when you are developing something. In this case, I am fixing a bug in ``tools.patch``, and it is very nice to be able to go to ``unittests/tools/files_patch_test`` and run all the relevant tests from there.

Comment (danimtb) [reply conans/test/unittests/tools/files_patch_test.py]: Having everything in the same place is not a valid reason and knowing what to run is easy in this case: everything that has a ``tools.patch``, although I agree we should have "functional" or "integration" test using the TestClient too.

For me this invalidates the reason or organizing the tests and will eventually evolve to the same mess.

**Pull request merged**